primitives/trie/src/cache.rs

```diff
@@ -0,0 +1,794 @@
+/// operations to the state. To not use all available memory it will ensure to stay in the
+/// bounds given via the [`Configuration`] at startup.
+///
+/// The instance of this object can be shared between multiple threads.
+pub struct SharedTrieCache<H: Hasher> {
+	node_cache: Arc<RwLock<SharedNodeCache<H>>>,
+	value_cache: Arc<RwLock<NoHashingLruCache<CachedValue<H::Out>>>>,
+	value_cache_element_size: usize,
+}
+
+impl<H: Hasher> Clone for SharedTrieCache<H> {
+	fn clone(&self) -> Self {
+		Self {
+			node_cache: self.node_cache.clone(),
+			value_cache: self.value_cache.clone(),
+			value_cache_element_size: self.value_cache_element_size,
+		}
+	}
+}
+
+impl<H: Hasher> SharedTrieCache<H> {
+	/// Create a new [`SharedTrieCache`].
+	pub fn new(config: Configuration) -> Self {
+		// The value cache element size isn't that huge, roughly around 50 bytes (mainly depending
+		// on the hash size). Thus, we only give it 5% of the overall cache size.
+		let value_cache_element_size =
+			mem::size_of::<u64>() + mem::size_of::<CachedValue<H::Out>>();
+		let value_cache_size_in_bytes = (config.maximum_size_in_bytes as f32 * 0.05) as usize;
+
+		Self {
+			node_cache: Arc::new(RwLock::new(SharedNodeCache::new(
+				config.maximum_size_in_bytes - value_cache_size_in_bytes,
+			))),
+			value_cache: Arc::new(RwLock::new(NoHashingLruCache::with_hasher(
+				value_cache_size_in_bytes / value_cache_element_size,
+				Default::default(),
+			))),
+			value_cache_element_size,
+		}
+	}
+
+	/// Create a new [`LocalTrieCache`] instance from this shared cache.
+	pub fn local_cache(&self) -> LocalTrieCache<H> {
+		LocalTrieCache {
+			shared: self.clone(),
+			node_cache: Default::default(),
+			value_cache: Default::default(),
+			shared_node_cache_access: Default::default(),
+			shared_value_cache_access: Default::default(),
+		}
+	}
+
+	/// Returns the used memory size of this cache in bytes.
+	pub fn used_memory_size(&self) -> usize {
+		let value_cache_size = self.value_cache.read().len() * self.value_cache_element_size;
+		let node_cache_size = self.node_cache.read().size_in_bytes;
+
+		node_cache_size + value_cache_size
+	}
+}
+
+/// The local trie cache.
+///
+/// This cache should be used per state instance created by the backend. One state instance is
+/// referring to the state of one block. It will cache all the accesses that are done to the state
+/// which could not be fullfilled by the [`SharedTrieCache`]. These locally cached items are merged
+/// back to the shared trie cache when this instance is dropped.
+///
+/// When using [`Self::as_trie_db_cache`] or [`Self::as_trie_db_mut_cache`], it will lock mutexes.
+/// So, it is important that these methods are not called multiple times, because they otherwise
+/// deadlock.
+pub struct LocalTrieCache<H: Hasher> {
+	/// The shared trie cache that created this instance.
+	shared: SharedTrieCache<H>,
+	/// The local cache for the trie nodes.
+	node_cache: Arc<Mutex<HashMap<H::Out, NodeOwned<H::Out>>>>,
```

Comment:
`node_cache` and `value_cache` should not really be behind an `Arc<Mutex<>>`.
I guess this is added because you want to have shallow backend clones, `duplicate` which is used in `wrap` to get a backend reference from an existing state, then create a recording wrapper around it. 
This design with stealing a backend from a state object and using it directly does not look right in the first place.
What you could do to make it a bit better and get rid of backend cloning is, instead of `as_trie_backend`, add a `into_trie_backend` function  that returns the backend by value. Then you could change `wrap` to take a value.